File: Algorithms/p033_Search_in_Rotated_Sorted_Array/p033_Search_in_Rotated_Sorted_Array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if(nums.empty()) return -1;
        int n = nums.size();
        int l = 0;
        int r = n;
        int left_most = nums[l];
        while(l<r){
            int mid = l+(r-l)/2;
            if(nums[mid]<left_most){
                r = mid;
            }else{
                l = mid+1;
            }
        }
        int p = l; // start of 2nd part, or n when no rotation
        if(target>=nums[0]&&target<=nums[p-1]){
            l = 0;
            r = p;
        }else if(p<n&&target>=nums[p]&&target<=nums[n-1]){
            l = p;
            r = n;
        }else{
            return -1;
        }
        while(l<r){
            int mid = l+(r-l)/2;
            if(target<nums[mid]){
                r = mid;
            }else if(target>nums[mid]){
                l = mid+1;
            }else{
                return mid;
            }
        }
        return -1;
    }
};

File: Algorithms/p033_Search_in_Rotated_Sorted_Array/p033_Search_in_Rotated_Sorted_Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums:
            return -1
        n = len(nums)
        l = 0
        r = n
        left_most = nums[l]
        while l<r:
            mid = l+(r-l)//2
            if nums[mid]<left_most:
                r = mid
            else:
                l = mid+1
        p = l # start of 2nd part, or n when no rotation
        if target>=nums[0] and target<=nums[p-1]:
            l = 0
            r = p
        elif p<n and target>=nums[p] and target<=nums[n-1]:
            l = p
            r = n
        else:
            return -1
        while l<r:
            mid = l+(r-l)//2
            if target<nums[mid]:
                r = mid
            elif target>nums[mid]:
                l = mid+1
            else:
                return mid
        return -1
